File: app/src/main/java/juego/chinchon/Activities/AcomodarActivity.java
package juego.chinchon.Activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import com.example.leoam.chinchonandroid.R;

import java.util.ArrayList;

import juego.chinchon.Constantes;
import juego.chinchon.Jugador;
import juego.chinchon.Mano;

/**
 *
 * @author LeoAM
 */
public class AcomodarActivity extends AppCompatActivity {

    // Colores de los indicadores de las cartas:
    private int clr_des;  // Carta no seleccionada
    private int clr_sel;  // Carta seleccionada
    private int clr_emp;  // Carta Emparejada

    // Estados de selección de las cartas:
    private int[] estados;  // Indica el estado de selección de cada carta.
    private static final int EST_DESELECCIONADO = 0;
    private static final int EST_SELECCIONADO = 1;
    private static final int EST_EMPAREJADO = 2;

    private ArrayList<Jugador> jugadores;
    private int jugadorActual;
    private int cortador;

    // Componentes de la pantalla:
    private LinearLayout filaBotones;
    private TableLayout tablaMano;
    private TextView errorTV, nombreTV, corteTV;

    private final View.OnClickListener cartaClickListener;

    public AcomodarActivity() {

        this.cartaClickListener = new View.OnClickListener() {

            public void onClick(View v) {
                int estaCarta = Integer.parseInt((String) v.getTag()) - 1;
                if (estaCarta != 8) {
                    View estado_carta = filaBotones.getChildAt(estaCarta);
                    switch (estados[estaCarta]) {
                        case EST_DESELECCIONADO:
                            estados[estaCarta] = EST_SELECCIONADO;
                            estado_carta.setBackgroundColor(clr_sel);
                            break;
                        case EST_SELECCIONADO:
                            estados[estaCarta] = EST_DESELECCIONADO;
                            estado_carta.setBackgroundColor(clr_des);
                            break;
                        default:
                    }
                }
            }
        };
        estados = new int[7];
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onCreate(Bundle icicle) {
        super.onCreate(icicle);
        setContentView(R.layout.acomodacion);

        jugadores = (ArrayList<Jugador>) getIntent().getSerializableExtra(Constantes.INTENT_JUGADORES);
        cortador = getIntent().getIntExtra(Constantes.INTENT_CORTE, 0);

        tablaMano = findViewById(R.id.ac_mano);
        TableRow fila;
        for (int f = 0; f < tablaMano.getChildCount(); f++) {
            fila = (TableRow) tablaMano.getChildAt(f);
            for (int c = 0; c < fila.getChildCount(); c++) {
                fila.getChildAt(c).setOnClickListener(cartaClickListener);
            }
        }

        filaBotones = findViewById(R.id.ac_buttonrow);

        final Button emparejar = findViewById(R.id.ac_emparejar_btn);
        final Button desarmar = findViewById(R.id.ac_desarmar_btn);
        final Button finalizar = findViewById(R.id.ac_finalizar_btn);

        nombreTV = findViewById(R.id.ac_tv_nombre);
        errorTV = findViewById(R.id.ac_errortext);
        corteTV = findViewById(R.id.ac_tv_corte);

        // Obtengo de R los colores para los indicadores:

        clr_des = ContextCompat.getColor(this, R.color.ac_deseleccionado);
        clr_sel = ContextCompat.getColor(this, R.color.ac_seleccionado);
        clr_emp = ContextCompat.getColor(this, R.color.ac_emparejado);

        emparejar.setOnClickListener(new View.OnClickListener() {

            public void onClick(View v) {
                ArrayList<Integer> tmpLista = new ArrayList<>();
                for (int carta = 0; carta < estados.length; carta++) {
                    if (estados[carta] == EST_SELECCIONADO) {
                        tmpLista.add(carta);
                    }
                }

                int[] indices = new int[tmpLista.size()];
                for (int carta = 0; carta < indices.length; carta++) {
                    indices[carta] = tmpLista.get(carta);
                }

                // Paso a comprobar si forman un juego:
                Mano mano = jugadores.get(jugadorActual).getMano();
                if (mano.mismoPalo(indices) || mano.mismoValor(indices)) {
                    for (int carta : tmpLista) {
                        estados[carta] = EST_EMPAREJADO;
                        filaBotones.getChildAt(carta).setBackgroundColor(clr_emp);
                    }
                    errorTV.setText(R.string.ac_acomodado);
                } else {
                    errorTV.setText(R.string.ac_noacomodado);
                }
            }

        });

        desarmar.setOnClickListener(new View.OnClickListener() {

            public void onClick(View v) {
                /* Se deseleccionan las cartas, colocando el estado en
                   "deseleccionado" y también se cambian el color de los
                   indicadores. */
                for (int carta = 0; carta < estados.length; carta++) {
                    estados[carta] = EST_DESELECCIONADO;
                    filaBotones.getChildAt(carta).setBackgroundColor(clr_des);
                }
            }

        });

        finalizar.setOnClickListener(new View.OnClickListener() {

            public void onClick(View v) {
                Jugador jugador = jugadores.get(jugadorActual);
                boolean[] acomodadas = new boolean[7];
                for (int carta = 0; carta < 7; carta++) {
                    acomodadas[carta] = (estados[carta] == EST_EMPAREJADO);
                }

                int puntos = jugador.getMano().getPuntos(acomodadas);

                if (jugadorActual == cortador) {
                    if (puntos == 0) {
                        /* Si el jugador corta y se queda sin cartas sin
                           acomodar entonces puede restar 10. */
                        jugador.restar10();
                    } else if (puntos > 5) {
                        /* Si el jugador corta y se quedó con más de 5 puntos en
                           la mano entonces se cancela el corte y se regresa al
                           mismo juego. */
                        Intent intent = new Intent();
                        setResult(2, intent);
                        finish();
                    } else {
                        jugador.addPuntos(puntos);
                    }
                } else {
                    jugador.addPuntos(puntos);
                }

                jugadorActual++;
                if (jugadorActual < jugadores.size()) {
                    // Cambio de jugador:
                    setJugadorEnPantalla();
                } else {
                    Intent intent = new Intent();
                    intent.putExtra(Constantes.INTENT_JUGADORES, jugadores);
                    setResult(1, intent);
                    finish();
                }
            }
        });

        // Empiezo la acomodación desde el primer jugador:
        jugadorActual = 0;

        /* Empiezo a configurar los views en pantalla, de acuerdo al jugador:
        También compruebo si el jugador hizo chinchón: */
        setJugadorEnPantalla();
    }

    private void setJugadorEnPantalla() {

        Jugador jugador = jugadores.get(jugadorActual);
        // Cambio el cuadro del nombre:
        nombreTV.setText(getString(R.string.ac_nombre, jugador.getNombre(), jugador.getPuntos()));

        //Indico que éste jugador fue el que cortó:
        if (cortador == jugadorActual) {
            corteTV.setText(R.string.ac_corte);
        } else {
            corteTV.setText("");
        }

        if (jugador.getMano().esChinchon() && (jugadorActual == cortador)) {
            // Si el jugador tiene chinchón entonces termina el juego:
            Intent intent = new Intent(AcomodarActivity.this, GanadorActivity.class);
            intent.putExtra(Constantes.INTENT_JUGADORES, jugadores);
            intent.putExtra(Constantes.INTENT_GANADOR, jugadorActual + 1);
            intent.putExtra(Constantes.INTENT_CHINCHON, true);
            startActivity(intent);
        } else {
            // Muestro la mano en pantalla:
            jugador.getMano().toTableLayout(tablaMano, false);

            // Pongo todas las banderas de las cartas en deseleccionado:
            for (int carta = 0; carta < 7; carta++) {
                estados[carta] = EST_DESELECCIONADO;
                filaBotones.getChildAt(carta).setBackgroundColor(clr_des);
            }
        }
        // Limpio el mensaje de error:
        errorTV.setText("");
    }
}

File: app/src/main/java/juego/chinchon/activities/AcomodarActivity.kt
package juego.chinchon.activities

import androidx.fragment.app.FragmentActivity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import com.example.leoam.chinchonkotlin.R
import juego.chinchon.Jugador
import juego.chinchon.Partida
import juego.chinchon.fragments.IManoFragment
import juego.chinchon.fragments.ManoFragment
import juego.chinchon.fragments.ManoFragment.Companion.EstadoSeleccion.*

import kotlinx.android.synthetic.main.acomodacion.*

/**
 *
 * @author LeoAM
 */
class AcomodarActivity : FragmentActivity(), IManoFragment {

    companion object {
        /** Cantidad de cartas en la mano. */
        private const val CANTIDAD_CARTAS: Int = 7
    }

    /** Índice del jugador actual. */
    private var jugadorActual = 0
    /**
     * Cantidad de cartas seleccionadas. Lo uso para determinar:
     * * El estado de habilitación del botón "Desarmar".
     * * El texto del botón "Cancelar"/"Finalizar", cuando el jugador actual
     * es el que cortó
     */
    private var cartasSeleccionadas = 0
        set(value) {
            field = value
            ac_desarmar_btn.isEnabled = cartasSeleccionadas != 0
            if (acomodaCortador()) {
                val texto = if (field == 0) { R.string.ac_Cancelar } else { R.string.ac_Finalizar }
                ac_finalizar_btn.setText(texto)
            }
        }
    /**
     * Cantidad de juegos formados hasta el momento. Solo puede haber un
     * máximo de dos juegos.
     */
    private var juegosActuales = 0
    private lateinit var partida: Partida
    private lateinit var manoFragment: ManoFragment

    /**
     * Inicializa la fase de acomodación de cartas.
     * * Inicializa los listeners de los botones.
     * * Pone al cortador como el primer jugador que puede acomodar sus cartas.
     * * Inicializa el fragmento de la mano.
     * * Calcula los puntos en total que se lleva el jugador por todas sus
     * cartas.
     */
    public override fun onCreate(icicle: Bundle?) {
        super.onCreate(icicle)
        setContentView(R.layout.acomodacion)

        partida = intent.getParcelableExtra("PARTIDA") as Partida

        ac_emparejar_btn.setOnClickListener(emparejarClickListener)
        ac_desarmar_btn.setOnClickListener(desarmarClickListener)
        ac_finalizar_btn.setOnClickListener(finalizarClickListener)

        jugadorActual = partida.rondaActual.cortador!!

        manoFragment = ManoFragment.newInstance()
        supportFragmentManager
                .beginTransaction()
                .replace(R.id.containerMano, manoFragment)
                .commit()

        calcularPuntos()
        setInformacionJugador()
    }

    /**
     * Muestra la mano del jugador que cortó. En el fragmento ya está
     * disponible la vista para ser modificada.
     */
    override fun onStart() {
        super.onStart()
        manoFragment.mostrarMano(partida.jugadores[jugadorActual].mano)
    }

    /**
     * Acción que se realiza al arrastrar una carta sobre otra en el fragmento.
     * Si se arrastra una carta sobre otra, entonces se cambian de lugar las
     * cartas. Si se arrastra una carta sobre si misma, o si solamente se hace
     * un toque sobre la carta, entonces la selecciona para tratar de formar un
     * juego.
     */
    override fun arrastrarCarta(origen: Int, destino: Int) {
        if (origen == destino) {
            if (origen in 0 until CANTIDAD_CARTAS) {
                when (manoFragment.getEstadoSeleccion(origen)) {
                    DESELECCIONADO -> {
                        manoFragment.seleccionarCarta(origen, SELECCIONADO)
                        cartasSeleccionadas++
                    }
                    SELECCIONADO -> {
                        manoFragment.seleccionarCarta(origen, DESELECCIONADO)
                        cartasSeleccionadas--
                    }
                    else -> {
                        // Hacer nada
                    }
                }
            }
        } else { // origen != destino
            val mano = partida.jugadores[jugadorActual].mano
            mano.swapCartas(origen, destino)
            manoFragment.mostrarMano(mano)

            val estadoSeleccionOrigen = manoFragment.getEstadoSeleccion(origen)
            val estadoSeleccionDestino = manoFragment.getEstadoSeleccion(destino)
            manoFragment.seleccionarCarta(origen, estadoSeleccionDestino)
            manoFragment.seleccionarCarta(destino, estadoSeleccionOrigen)
        }
    }

    /**
     * Click listener del botón "Emparejar".
     * Genera una lista de indices a partir de las cartas con estado
     * "seleccionado" y las pasa al objeto mano del jugador actual para
     * determinar si las cartas seleccionadas forman un juego. Si es asi,
     * cambia el estado de las cartas a "emparejado" y actualiza el cuadro de
     * texto con los puntos del jugador. Si no, muestra un error.
     */
    private val emparejarClickListener: View.OnClickListener = View.OnClickListener {
        val tmpListaIndices = ArrayList<Int>()
        for (i in 0 until CANTIDAD_CARTAS) {
            val estado = manoFragment.getEstadoSeleccion(i)
            if (estado == SELECCIONADO) {
                tmpListaIndices.add(i)
            }
        }

        val indices = tmpListaIndices.toIntArray()
        if (partida.rondaActual.formanJuego(jugadorActual, indices)) {
            val estadoSeleccionJuego = when(juegosActuales) {
                0 -> JUEGO_1
                1 -> JUEGO_2
                else -> {
                    throw IllegalStateException("Hay más de 2 juegos")
                }
            }
            juegosActuales++
            for (i in indices) {
                manoFragment.seleccionarCarta(i, estadoSeleccionJuego)
            }
            Toast.makeText(this, getText(R.string.ac_acomodado), Toast.LENGTH_SHORT).show()
            calcularPuntos()
        } else {
            Toast.makeText(this, getText(R.string.ac_noacomodado), Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * Click listener del botón "Desarmar".
     * Cambia el estado de todas las cartas a "deseleccionado", desarmando los
     * juegos que se hallan formado. Actualiza el textview con los puntos del
     * jugador.
     */
    private val desarmarClickListener: View.OnClickListener = View.OnClickListener {
        manoFragment.limpiarSeleccion()
        juegosActuales = 0
        calcularPuntos()
        cartasSeleccionadas = 0
    }

    /**
     * Click listener del botón "Finalizar". Genera una lista de booleanos con
     * las cartas que fueron emparejadas y las pasa al objeto mano para
     * calcular cuantos puntos hizo. Si es el cortador e hizo más de 5 puntos,
     * entonces vuelve a la partida con un mensaje de error.
     */
    private val finalizarClickListener: View.OnClickListener = View.OnClickListener {
        try {
            if (partida.acomodar(jugadorActual, cartasEmparejadas())) {
                // Acomodó el que cortó, sigo con el otro jugador
                jugadorActual = 1 - jugadorActual
                setJugadorEnPantalla()
                calcularPuntos()
            } else {
                val intent = Intent()
                intent.putExtra("PARTIDA", partida)
                setResult(1, intent)
                finish()
            }
        } catch (e: IllegalStateException) {
            val intent = Intent()
            setResult(2, intent)
            finish()
        }
    }

    /**
     * Determina si el jugador que está acomodando las cartas es el que cortó
     * en esta ronda.
     */
    private fun acomodaCortador(): Boolean {
        return jugadorActual == partida.rondaActual.cortador
    }

    /**
     * Muestra la información del jugador y sus cartas en la pantalla.
     * * Indica si el jugador es el que cortó en la última ronda.
     * * Configura el estado de selección de las cartas para que no esté
     * ninguna seleccionada al inicio.
     */
    private fun setJugadorEnPantalla() {
        juegosActuales = 0
        setInformacionJugador()
        manoFragment.mostrarMano(partida.jugadores[jugadorActual].mano)
        manoFragment.limpiarSeleccion()
        cartasSeleccionadas = 0
    }

    /**
     * Calcula los puntos que va a sumar el jugador y los puntos que va a tener
     * en total si acomoda las cartas. Lo muestra en un TextView.
     */
    private fun calcularPuntos() {
        val jugador: Jugador = partida.jugadores[jugadorActual]
        val puntosAhora = jugador.puntos
        val puntosTurno = jugador.mano.getPuntos(cartasEmparejadas())

        if (acomodaCortador() && puntosTurno == 0) {
            ac_tv_puntos.text = getString(R.string.ac_restapuntos, puntosAhora - 10)
        } else {
            ac_tv_puntos.text = getString(R.string.ac_sumapuntos, puntosTurno, puntosAhora + puntosTurno)
        }
    }

    /**
     * Muestra la información del jugador en la pantalla:
     * * Nombre del jugador y los puntos que tuvo hasta la ronda pasada.
     * * Si fue el que cortó en esta ronda.
     */
    private fun setInformacionJugador() {
        val jugador = partida.jugadores[jugadorActual]

        ac_tv_nombre.text = getString(R.string.ac_nombre, jugador.nombre, jugador.puntos)

        if (acomodaCortador()) {
            ac_tv_corte.setText(R.string.ac_corte)
        } else {
            ac_tv_corte.text = ""
        }
    }

    /**
     * Devuelve un array booleano donde cada `true` representa una carta que
     * forma parte de un juego (está acomodada).
     */
    private fun cartasEmparejadas(): BooleanArray {
        return BooleanArray(CANTIDAD_CARTAS) { i ->
            manoFragment.getEstadoSeleccion(i) == JUEGO_1 ||
            manoFragment.getEstadoSeleccion(i) == JUEGO_2
        }
    }
}
